// @mugen/shared/src/engines/ability/index.ts
import type { UnitInstance, Position } from '../../types/index.js';
import { AbilityType, CombatModifierType } from '../../types/index.js';
import type { Result } from '../../types/actions.js';

export interface AbilityResult {
  unit: UnitInstance;
  target: UnitInstance | null;
  lifeCost: number;
}

const ABILITY_DAMAGE_AMOUNT = 3;
const ABILITY_HEAL_AMOUNT = 3;

export function useAbility(
  unit: UnitInstance,
  target: UnitInstance | null
): Result<AbilityResult> {
  if (unit.hasUsedAbilityThisTurn) {
    return { ok: false, error: 'Ability already used this turn' };
  }

  const ability = unit.card.ability;

  const updatedUnit: UnitInstance = {
    ...unit,
    hasUsedAbilityThisTurn: true,
  };

  let updatedTarget = target;

  switch (ability.abilityType) {
    case AbilityType.DAMAGE: {
      if (target) {
        const newHp = Math.max(0, target.currentHp - ABILITY_DAMAGE_AMOUNT);
        updatedTarget = { ...target, currentHp: newHp };
      }
      break;
    }
    case AbilityType.HEAL: {
      if (target) {
        const maxHp = target.card.maxHp;
        const newHp = Math.min(maxHp, target.currentHp + ABILITY_HEAL_AMOUNT);
        updatedTarget = { ...target, currentHp: newHp };
      }
      break;
    }
    case AbilityType.BUFF: {
      if (target) {
        updatedTarget = {
          ...target,
          combatModifiers: [
            ...target.combatModifiers,
            { type: CombatModifierType.ATK_BUFF, duration: 1 },
          ],
        };
      }
      break;
    }
    case AbilityType.MODIFIER: {
      if (target) {
        updatedTarget = {
          ...target,
          combatModifiers: [
            ...target.combatModifiers,
            { type: CombatModifierType.NO_COUNTERATTACK, duration: 1 },
          ],
        };
      }
      break;
    }
  }

  return {
    ok: true,
    value: {
      unit: updatedUnit,
      target: updatedTarget,
      lifeCost: 0,
    },
  };
}

export function resetAbilityUsage(units: UnitInstance[]): UnitInstance[] {
  return units.map((u) => ({ ...u, hasUsedAbilityThisTurn: false }));
}

export function hasUsedAbility(unit: UnitInstance): boolean {
  return unit.hasUsedAbilityThisTurn;
}

function chebyshevDistance(a: Position, b: Position): number {
  return Math.max(Math.abs(a.x - b.x), Math.abs(a.y - b.y));
}

export interface AbilityTarget {
  unitId: string;
  ownerId: string;
  position: Position;
}

export function getAbilityTargets(
  unit: UnitInstance,
  allUnits: UnitInstance[],
  ownerId: string
): AbilityTarget[] {
  if (!unit.position) return [];
  if (unit.hasUsedAbilityThisTurn) return [];

  const abilityType = unit.card.ability.abilityType;
  const range = unit.card.range;

  const targets: AbilityTarget[] = [];

  for (const candidate of allUnits) {
    if (!candidate.position) continue;
    if (candidate.currentHp <= 0) continue;

    const dist = chebyshevDistance(unit.position, candidate.position);
    if (dist > range) continue;

    const isFriendly = candidate.ownerId === ownerId;

    switch (abilityType) {
      case AbilityType.DAMAGE:
      case AbilityType.MODIFIER:
        if (!isFriendly) {
          targets.push({ unitId: candidate.card.id, ownerId: candidate.ownerId, position: candidate.position });
        }
        break;
      case AbilityType.HEAL:
      case AbilityType.BUFF:
        if (isFriendly) {
          targets.push({ unitId: candidate.card.id, ownerId: candidate.ownerId, position: candidate.position });
        }
        break;
    }
  }

  return targets;
}
